compiler/passes/cleanup.cpp

```diff
@@ -206,6 +206,20 @@ static void add_parens_to_deinit_fns(FnSymbol* fn) {
 }
 
 
+/* Change the type of the receiver in type methods from "_any" to
+   the type that defined the method.
+*/
+static void set_receiver_type_in_type_methods(FnSymbol* fn) {
+  if (fn->hasFlag(FLAG_METHOD) && fn->thisTag == INTENT_TYPE) {
+    if (TypeSymbol* ts = toTypeSymbol(fn->defPoint->parentSymbol)) {
+      if (fn->_this->type == dtAny) {
+        fn->_this->type = ts->type;
+      }
```

Comment:
Secondary methods fall through here, and basically the same transformation happens to them in normalization, though it wasn't super obvious exactly where in normalization it happens.